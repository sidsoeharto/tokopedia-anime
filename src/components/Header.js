/** @jsxImportSource @emotion/react */

import React from "react";
import { solid } from "@fortawesome/fontawesome-svg-core/import.macro";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useLocation, useNavigate, matchPath } from "react-router-dom";
import AppColors from "../styles/AppColors";

const Header = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const newLocation = { ...location };

  const showBackIcon = location.pathname !== "/";
  const showEditButton = !!matchPath({
    path: "/collections/:name",
    exact: true,
    strict: false
  }, location.pathname);

  return (
    <div css={{ position: "sticky", top: 0, zIndex: 3}}>
      <div
        css={{
          height: "5vh",
          padding: "0.5rem 2rem",
          backgroundColor: AppColors.green600,
          display: "flex",
          alignItems: "center",
          justifyContent: "space-between",
          filter: "drop-shadow(0 4px 3px rgb(0 0 0 / 0.07)) drop-shadow(0 2px 2px rgb(0 0 0 / 0.06))",
        }}
      >
        <div
          css={{
            background: 'none',
            border: 'none',
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
          }}
        >
          {showBackIcon && (
            <FontAwesomeIcon
              icon={solid("arrow-left")}
              css={{ 
                marginRight: 12 ,
                color: AppColors.gray100,
                display: 'inline-flex',
                cursor: 'pointer',
              }}
              onClick={() => navigate(-1)}
            />
          )}
          <span
            css={{
              marginTop: 2,
              fontSize: 18,
              fontWeight: "bolder",
              display: "inline-flex",
              alignItems: 'center',
              color: AppColors.gray100,
            }}
          >
            {newLocation.state?.title ?? "AnimeList"}
          </span>
        </div>
        
        {!showBackIcon && (
          <button
            css={{
              background: 'none',
              border: 'none',
              display: 'flex',
              flexDirection: 'row',
              cursor: 'pointer',
            }}
            onClick={() =>
              navigate("/collections", { state: { title: "Collection List" } })
            }
          >
            <FontAwesomeIcon
              icon={solid("heart")}
              css={{ 
                marginLeft: 12, 
                marginRight: 2, 
                display: "inline-block",
              }}
              size="2x"
              color={AppColors.pink300}
            />
          </button>
        )}
        {
          showEditButton && (
            <button
              css={{
                background: AppColors.pink400,
                color: AppColors.gray100,
                borderRadius: 6,
                padding: '0.625rem 1.25rem',
                border: 'none',
                display: 'flex',
                flexDirection: 'row',
                cursor: 'pointer',
                fontSize: 16,
                alignItems: 'center',
                fontFamily: 'Overpass',
              }}
            >
              <FontAwesomeIcon
                icon={solid("edit")}
                css={{  
                  marginRight: 8, 
                  display: "inline-flex",
                }}
                size="1x"
                color={AppColors.gray100}
              />
              <span 
                css={{marginTop: 3}}
              >
                Edit Collection
              </span>
            </button>
          )
        }
      </div>
    </div>
  );
};

export default Header;