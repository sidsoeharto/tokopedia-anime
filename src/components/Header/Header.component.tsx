/** @jsxImportSource @emotion/react */

import { solid } from "@fortawesome/fontawesome-svg-core/import.macro";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useLocation, useNavigate } from "react-router-dom";
import AppColors from "../../styles/AppColors";
import {
  headerContainer,
  headerSubContainer,
  backIcon,
  titleContainer,
  title,
  collectionContainer,
  collectionIcon,
} from "./Header.styles";

const Header = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const newLocation = { ...location };

  const showBackIcon = location.pathname !== "/";

  return (
    <div css={headerContainer}>
      <div css={headerSubContainer}>
        <div css={titleContainer}>
          {showBackIcon && (
            <button
              style={{ all: 'unset' }}
              onClick={() => navigate(-1)}
              data-testid="back-icon"
            >
              <FontAwesomeIcon
                icon={solid("arrow-left")}
                css={backIcon}
              />
            </button>
          )}
          <span css={title}>{newLocation.state?.title ?? "AnimeList"}</span>
        </div>

        {!showBackIcon && (
          <button
            css={collectionContainer}
            onClick={() =>
              navigate("/collections", { state: { title: "Collection List" } })
            }
            data-testid="back-icon"
          >
            <FontAwesomeIcon
              icon={solid("heart")}
              css={collectionIcon}
              size="lg"
              color={AppColors.pink200}
            />
            Collections
          </button>
        )}
      </div>
    </div>
  );
};

export default Header;
