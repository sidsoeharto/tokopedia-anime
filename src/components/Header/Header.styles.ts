import { css } from "@emotion/react";

import { mediaQueryMaxWidth } from "../../styles/config";
import AppColors from "../../styles/AppColors";

export const headerContainer = css`
  position: sticky;
  top: 0;
  z-index: 3;
`;

export const headerSubContainer = css`
  height: 5vh;
  padding: 0.5rem 2rem;
  background-color: ${AppColors.green600};
  display: flex;
  align-items: center;
  justify-content: space-between;
  filter: drop-shadow(0 4px 3px rgba(0, 0, 0, 0.07))
    drop-shadow(0 2px 2px rgba(0, 0, 0, 0.06));
`;

export const backIcon = css`
  margin-right: 12px;
  color: ${AppColors.gray100};
  display: inline-flex;
  cursor: pointer;
`;

export const titleContainer = css`
  background: none;
  border: none;
  display: flex;
  flex-direction: row;
  align-items: center;
`;

export const title = css`
  margin-top: 2px;
  font-size: 18px;
  font-weight: bolder;
  display: inline-flex;
  align-items: center;
  color: ${AppColors.gray100};
`;

export const collectionContainer = css`
  background: ${AppColors.pink500};
  padding: 0.3rem 0.8rem;
  border-radius: 12px;
  border: none;
  display: inline-flex;
  align-items: center;
  color: ${AppColors.pink100};
  font-family: Overpass, sans-serif;
  font-size: 0.9rem;
  flex-direction: row;
  cursor: pointer;
  ${mediaQueryMaxWidth[1]} {
    font-size: 0.8rem;
    padding: 4px 8px;
  }
`;

export const collectionIcon = css`
  margin-right: 6px;
  display: inline-block;
`;
