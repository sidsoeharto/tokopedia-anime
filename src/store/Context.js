import React, { createContext, useEffect, useReducer } from "react";
import { useLocalStorage } from "../hooks/usePersist";
import reducer, { initialState } from "./reducer";

function Context() {
  const AppContext = createContext(initialState);

  const Provider = (props) => {
    const [value, setValue] = useLocalStorage("globalState");
    const initialStore = Object.keys(value).length > 0 ? value : initialState;
    const store = useReducer(reducer, initialStore);

    useEffect(() => {
      setValue(store[0]);
    }, [setValue, store, value]);

    return (
      <AppContext.Provider value={store}>{props.children}</AppContext.Provider>
    );
  };

  const Consumer = AppContext.Consumer;
  
  return {
    Provider,
    Consumer,
    AppContext,
  };
}

export default Context();