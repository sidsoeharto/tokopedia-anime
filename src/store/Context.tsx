import React, { createContext, useEffect, useReducer, ReactNode } from "react";
import { useLocalStorage } from "../hooks/usePersist";
import reducer, { initialState, State } from "./reducer";
import { Action } from "./actions";

interface ContextProps {
  Provider: React.FC<{ children: ReactNode }>;
  Consumer: React.Consumer<[State, React.Dispatch<Action>]>;
  AppContext: React.Context<[State, React.Dispatch<Action>]>;
}

const AppContext = createContext<[State, React.Dispatch<Action>]>([initialState, () => {}]);

const Provider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [value, setValue] = useLocalStorage("globalState");
  const initialStore = Object.keys(value)?.length > 0 ? value : initialState;
  const store = useReducer(reducer, initialStore);

  useEffect(() => {
    setValue(store[0]);
  }, [setValue, store, value]);

  return (
    <AppContext.Provider value={store}>{children}</AppContext.Provider>
  );
};

const Consumer = AppContext.Consumer;

const Context: ContextProps = {
  Provider,
  Consumer,
  AppContext,
};

export default Context;
