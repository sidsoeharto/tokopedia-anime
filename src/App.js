/** @jsxImportSource @emotion/react */

import React from "react";
import Routes from "./routes/route";
import { BrowserRouter } from "react-router-dom";
import Context from "./store/Context";
import {Header} from "./components";
import AppColors from "./styles/AppColors";

import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
} from "@apollo/client";

const client = new ApolloClient({
  uri: 'https://graphql.anilist.co/',
  cache: new InMemoryCache()
});


function App() {
  return (
    <ApolloProvider client={client}>
      <Context.Provider>
        <BrowserRouter>
          <div css={{backgroundColor: AppColors.gray100}}>
            <div
              css={{
                margin: "0 auto",
                display: "block",
                backgroundColor: "white",
              }}
            >
              <Header />
              <Routes />
            </div>
          </div>
        </BrowserRouter>
      </Context.Provider>
    </ApolloProvider>
  );
}

export default App;