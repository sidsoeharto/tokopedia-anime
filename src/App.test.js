import { createRoot } from 'react-dom/client';
import App from './App';
import renderer from "react-test-renderer";
import { render, cleanup, screen } from "@testing-library/react";

import "@testing-library/jest-dom/extend-expect";
import AppRouter from "./routes/route";
import { BrowserRouter } from "react-router-dom";
import Context from './store/Context';

import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
} from "@apollo/client";

import Header from './components/Header';

afterEach(cleanup);

test('renders app', () => {
  const div = document.createElement('div');
  const root = createRoot(div);
  root.render(<App />);
  root.unmount(div);
});

describe('Render Animelist Application', () => {
  const client = new ApolloClient({
    uri: 'https://graphql.anilist.co/',
    cache: new InMemoryCache()
  });

  it('should render initial page', () => {
    const tree = renderer.create(
      <ApolloProvider client={client}>
        <Context.Provider>
          <BrowserRouter>
            <Header />
            <AppRouter />
          </BrowserRouter>
        </Context.Provider>
      </ApolloProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  })
})

