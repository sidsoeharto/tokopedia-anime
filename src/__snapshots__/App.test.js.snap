// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Animelist Application should render initial page 1`] = `
Array [
  <div
    className="css-2nviu7-Header"
  >
    <div
      className="css-bnr0wc-Header"
    >
      <div
        className="css-yfzaqb-Header"
      >
        <span
          className="css-1eb1ylz-Header"
        >
          AnimeList
        </span>
      </div>
      <button
        className="css-1fcbqhf-Header"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-heart fa-2x css-1tkumux-Header"
          color="#f9a8d4"
          data-icon="heart"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84.02L256 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 .0003 232.4 .0003 190.9L0 190.9z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>,
  <div
    className="css-12s9sh8-Loading"
  >
    <div
      className="css-ep5w2r-Loading"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-spinner fa-2x css-1io8az8-Loading"
        color="#a1a1aa"
        data-icon="spinner"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48C304 74.51 282.5 96 256 96C229.5 96 208 74.51 208 48C208 21.49 229.5 0 256 0C282.5 0 304 21.49 304 48zM304 464C304 490.5 282.5 512 256 512C229.5 512 208 490.5 208 464C208 437.5 229.5 416 256 416C282.5 416 304 437.5 304 464zM0 256C0 229.5 21.49 208 48 208C74.51 208 96 229.5 96 256C96 282.5 74.51 304 48 304C21.49 304 0 282.5 0 256zM512 256C512 282.5 490.5 304 464 304C437.5 304 416 282.5 416 256C416 229.5 437.5 208 464 208C490.5 208 512 229.5 512 256zM74.98 437C56.23 418.3 56.23 387.9 74.98 369.1C93.73 350.4 124.1 350.4 142.9 369.1C161.6 387.9 161.6 418.3 142.9 437C124.1 455.8 93.73 455.8 74.98 437V437zM142.9 142.9C124.1 161.6 93.73 161.6 74.98 142.9C56.24 124.1 56.24 93.73 74.98 74.98C93.73 56.23 124.1 56.23 142.9 74.98C161.6 93.73 161.6 124.1 142.9 142.9zM369.1 369.1C387.9 350.4 418.3 350.4 437 369.1C455.8 387.9 455.8 418.3 437 437C418.3 455.8 387.9 455.8 369.1 437C350.4 418.3 350.4 387.9 369.1 369.1V369.1z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <p
        className="css-1khqfiv-Loading"
      >
        Loading...
      </p>
    </div>
  </div>,
]
`;
